### PR TITLE
Core policy

### PR DESCRIPTION
**Pull request by kraftp**

### CODE REVIEW

File: src/CorePolicy.h [hunk 1]
```diff
@@ -0,0 +1,85 @@
+/* Copyright (c) 2015-2017 Stanford University
```
Comment (johnousterhout) [inline src/CorePolicy.h]: Copyright (c) 2017 ...

(should reflect date of first version of file; fix CorPolicy.cc also)

File: src/CorePolicy.h [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+#include <string>
+#include <thread>
+
+#include "PerfUtils/Cycles.h"
+#include "PerfUtils/Util.h"
+#include "Logger.h"
+#include "PerfStats.h"
+#include "Common.h"
+
+#define MAX_THREAD_CLASSES 64
```
Comment (johnousterhout) [inline src/CorePolicy.h]: A definition like this needs a comment. For example, why are you limiting the number of thread classes?

File: src/CorePolicy.h [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+
+typedef uint32_t threadClass_t;
```
Comment (johnousterhout) [inline src/CorePolicy.h]: In our coding style we don't use "_t" notation, but it should be capitalized: ThreadClass

File: src/CorePolicy.h [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+
+/** Struct used to define the threadCoreMap in CorePolicy. **/
```
Comment (johnousterhout) [inline src/CorePolicy.h]: Too vague. What is this thing? I should be able to read this comment and get a decent feel for what this struct does.

File: src/CorePolicy.h [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+struct threadCoreMapEntry {
+  uint32_t numFilled;
```
Comment (johnousterhout) [inline src/CorePolicy.h]: Every member variable in every struct/class should have a comment.

File: src/CorePolicy.h [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+#include "Logger.h"
+#include "PerfStats.h"
+#include "Common.h"
+
+#define MAX_THREAD_CLASSES 64
+
+typedef uint32_t threadClass_t;
+
+/** Struct used to define the threadCoreMap in CorePolicy. **/
+struct threadCoreMapEntry {
```
Comment (johnousterhout) [inline src/CorePolicy.h]: Class/struct names are always capitalized: ThreadCoreMapEntry

File: src/CorePolicy.h [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+  uint32_t numFilled;
+  int* map;
+};
+
+class CorePolicy {
```
Comment (johnousterhout) [inline src/CorePolicy.h]: Need an overall comment describing the basic abstraction provided by this class. Every class/struct must always have a header comment.

File: src/CorePolicy.h [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+  public:
+    /** Constructor and destructor for CorePolicy. */
```
Comment (johnousterhout) [inline src/CorePolicy.h]: Why are the bodies of these methods here in the header file, rather than in CorePolicy.cc. In general, method bodies should only be in the header file if that's necessary, either because they are templated, or because we need them to be inline for performance.

Also, both of these methods need header comments (which should be wherever the method bodies are located).

File: src/CorePolicy.h [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+    threadCoreMapEntry* getThreadCoreMapEntry(threadClass_t threadClass);
+
+    threadClass_t baseClass = 0;
+    
+  protected:
+    /**
+      * A map from thread classes to cores on which threads of those classes
+      * can run.  If threadCoreMap[i]->map[j] = c for some
+      * j < threadCoreMap[i]->numFilled then Arachne can create a thread of
+      * class i on the core with coreId c.
```
Comment (johnousterhout) [inline src/CorePolicy.h]: Most of this documentation belongs with the ThreadCoreMapEntry class, not here.

It appears that storage for this is dynamically allocated? If so, that definitely needs to be documented here. However, do you need that? Couldn't you just use an STL vector and let C++ handle that? Let's discuss.

File: src/CorePolicy.h [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+struct threadCoreMapEntry {
+  uint32_t numFilled;
+  int* map;
+};
+
+class CorePolicy {
+  public:
+    /** Constructor and destructor for CorePolicy. */
+    CorePolicy() {
+        threadCoreMap = (threadCoreMapEntry**) malloc(MAX_THREAD_CLASSES * sizeof(threadCoreMapEntry*));
```
Comment (johnousterhout) [inline src/CorePolicy.h]: Is this line longer than 80 columns? If so, wrap it.

File: src/CorePolicy.cc [hunk 11]
```diff
@@ -0,0 +1,201 @@
...
+/**
+  * Arachne will attempt to increase the number of cores if the idle core
+  * fraction (computed as idlePercentage * numSharedCores) is less than this
+  * number.
+  */
+double maxIdleCoreFraction = 0.1;
+
+/**
+  * Arachne will attempt to increase the number of cores if the load factor
+  * increase beyond this threshold.
```
Comment (johnousterhout) [inline src/CorePolicy.cc]: increase => increases

File: src/CorePolicy.cc [hunk 12]
```diff
@@ -0,0 +1,201 @@
...
+double loadFactorThreshold = 1.0;
+
+/**
+  * Only here to satisfy compiling benchmark.
+  */
+double maxUtilization = 0.9;
+
+/**
+  * Save the core fraction at which we ramped up based on load factor, so we
+  * can decide whether to ramp down.
```
Comment (johnousterhout) [inline src/CorePolicy.cc]: Looks like this is malloced? If so, that needs to be documented. Better yet, switch to std::vector<double>.

File: src/CorePolicy.cc [hunk 13]
```diff
@@ -0,0 +1,201 @@
...
+  */
+double *utilizationThresholds;
+
+/**
+  * The difference in load, expressed as a fraction of a core, between a
+  * ramp-down threshold and the corresponding ramp-up threshold.
```
Comment (johnousterhout) [inline src/CorePolicy.cc]: threshold. => threshold (i.e., we wait to ramp down until the load gets a bit below the point at which we ramped up).

File: src/CorePolicy.cc [hunk 14]
```diff
@@ -0,0 +1,201 @@
...
+  */
+double idleCoreFractionHysteresis = 0.2;
+
+/**
+  * The percentage of slots which are occupied that will prevent reducing the
```
Comment (johnousterhout) [inline src/CorePolicy.cc]: What slots?

Also, grammatically incorrect, and not clear which way the inequality works. How about this, and instead?

If at least this fraction of <what kind of?> slots are occupied, we will not reduce the number of cores.

File: src/CorePolicy.cc [hunk 15]
```diff
@@ -0,0 +1,201 @@
...
+const uint64_t MEASUREMENT_PERIOD = 50 * 1000 * 1000;
+
+void coreLoadEstimator();
+
+/**
+  * Bootstrap the core load estimator thread.  This must be done separately
+  * from the CorePolicy constructor because the constructor must run before
+  * Arachne adds any cores, but the core load estimator must run after
+  * Arachne has cores.
+  */
```
Comment (johnousterhout) [inline src/CorePolicy.cc]: I'm still not wild about this method; it feels like an awkward addition to the API. I think there are ways to achieve the desired effect without an additional exported method. Let's discuss.

File: src/CorePolicy.cc [hunk 16]
```diff
@@ -0,0 +1,201 @@
...
+void CorePolicy::bootstrapLoadEstimator(bool disableLoadEstimation) {
```
Comment (johnousterhout) [inline src/CorePolicy.cc]: What's the purpose of the argument? Why even bother to call this method if it's true? Of course, if I get my way above, the whole method will go away…

File: src/CorePolicy.cc [hunk 17]
```diff
@@ -0,0 +1,201 @@
...
+    }
+}
+
+/**
+  * Choose a core to deschedule.  Return its coreId.
+  */
+int CorePolicy::chooseRemovableCore() {
+    threadCoreMapEntry* entry = threadCoreMap[baseClass];
+    uint8_t minLoaded =
+        Arachne::occupiedAndCount[entry->map[0]]->load().numOccupied;
```
Comment (johnousterhout) [inline src/CorePolicy.cc]: Major information leakage! I'd really rather not have Arachne exporting its occupiedAndCount vector. Let's discuss.

File: src/CorePolicy.h [hunk 18]
```diff
@@ -0,0 +1,85 @@
...
+        free(threadCoreMap);
+    }
+    virtual void bootstrapLoadEstimator(bool disableLoadEstimation);
+    virtual int chooseRemovableCore();
+    virtual void addCore(int coreId);
+    virtual void removeCore(int coreId);
+    virtual uint32_t maxClass() { return 0U; }
+    threadCoreMapEntry* getThreadCoreMapEntry(threadClass_t threadClass);
+
+    threadClass_t baseClass = 0;
```
Comment (johnousterhout) [inline src/CorePolicy.h]: Needs documentation.

File: src/CorePolicy.cc [hunk 19]
```diff
@@ -0,0 +1,201 @@
...
+        utilizationThresholds = new double[Arachne::maxNumCores];
+        Arachne::createThread(baseClass, coreLoadEstimator);
+    }
+}
+
+/**
+  * Choose a core to deschedule.  Return its coreId.
+  */
+int CorePolicy::chooseRemovableCore() {
+    threadCoreMapEntry* entry = threadCoreMap[baseClass];
```
Comment (johnousterhout) [inline src/CorePolicy.cc]: Why do you use only baseClass for this decision? Let's discuss.

File: src/CorePolicy.cc [hunk 20]
```diff
@@ -0,0 +1,201 @@
...
+
+/**
+  * Update threadCoreMap when a core is removed.  Takes in the coreId
+  * of the doomed core.
+**/
+void CorePolicy::removeCore(int coreId) {
+  threadCoreMapEntry* entry = threadCoreMap[baseClass];
+  entry->numFilled--;
+  int saveCoreId = entry->map[entry->numFilled];
+  for (uint32_t i = 0; i < entry->numFilled; i++) {
```
Comment (johnousterhout) [inline src/CorePolicy.cc]: Are there concurrency issues if new threads are created while threadCoreMap is being updated? Let's discuss. If nothing else, concurrency needs to be mentioned in the documentation for this variable (I suspect you have followed certain constraints in your coding, but I don't see any place where those constraints are documented).

File: src/CorePolicy.cc [hunk 21]
```diff
@@ -0,0 +1,201 @@
...
+}
+
+threadCoreMapEntry* CorePolicy::getThreadCoreMapEntry(threadClass_t threadClass) {
+  return threadCoreMap[threadClass];
+}
+
+/**
+  * Periodically wake up and observe the current load in Arachne to determine
+  * whether it is necessary to increase or reduce the number of cores used by
+  * Arachne.
```
Comment (johnousterhout) [inline src/CorePolicy.cc]: Mention that this method runs as the top-level method in a thread.

File: src/CorePolicy.cc [hunk 22]
```diff
@@ -0,0 +1,201 @@
...
+  */
+void coreLoadEstimator() {
+    Arachne::PerfStats previousStats;
+    Arachne::PerfStats::collectStats(&previousStats);
+
+    for (Arachne::PerfStats currentStats; ; previousStats = currentStats) {
```
Comment (johnousterhout) [inline src/CorePolicy.cc]: What does one iteration of this loop correspond to? That would make for a good comment just above the loop.

File: src/CorePolicy.cc [hunk 23]
```diff
@@ -0,0 +1,201 @@
...
+        Arachne::sleep(MEASUREMENT_PERIOD);
+        Arachne::PerfStats::collectStats(&currentStats);
+
+        // Take a snapshot of currently active cores before performing
+        // estimation to avoid races between estimation and the fulfillment of
+        // a previous core request.
+        uint32_t curActiveCores = Arachne::numActiveCores;
```
Comment (johnousterhout) [inline src/CorePolicy.cc]: Another leak of data that feels like it should be private to Arachne. Let's discuss.

File: src/CorePolicy.cc [hunk 24]
```diff
@@ -0,0 +1,201 @@
...
+
+        // Exclusive cores should contribute nothing to statistics relevant to
+        // core estimation during the period over which they are exclusive.
+        int numSharedCores = curActiveCores - Arachne::numExclusiveCores;
```
Comment (johnousterhout) [inline src/CorePolicy.cc]: ?? exclusive cores ?? Let's discuss.

File: src/Arachne.cc [hunk 25]
```diff
@@ -910,20 +868,21 @@ ThreadContext::initializeStack() {
  *    remove the options that Arachne recognizes.
  */
 void
-init(int* argcp, const char** argv) {
+init(CorePolicy* initCorePolicy, int* argcp, const char** argv) {
```
Comment (johnousterhout) [inline src/Arachne.cc]: Be sure to add a new \param comment above when you add a new argument.

File: src/Arachne.cc [hunk 26]
```diff
@@ -1245,10 +1192,10 @@ void descheduleCore() {
     // hold it for too long.
     // If this creation fails, it would implies that we are overloaded and
     // should not ramp down.
-    if (createThreadOnCore(minIndex, releaseCore) == NullThread) {
+    if (createThreadOnCore(corePolicy->baseClass, minCoreId, releaseCore) == NullThread) {
```
Comment (johnousterhout) [inline src/Arachne.cc]: Information leakage: why is Arachne referencing corePolicy->baseClass? Let's discuss.

File: src/Arachne.cc [hunk 27]
```diff
@@ -220,6 +177,8 @@ bool useCoreArbiter = true;
 
 CoreArbiterClient* coreArbiter = NULL;
 
+CorePolicy* corePolicy = NULL;
```
Comment (johnousterhout) [inline src/Arachne.cc]: Needs documentation.

File: src/Arachne.cc [hunk 28]
```diff
@@ -1358,18 +1305,30 @@ bool makeExclusiveOnCore(bool forScaleDown) {
...
+    int nextMigrationTarget = 0;
 
     for (uint8_t i = 0; i < maxThreadsPerCore; i++) {
         if (i == originalContext->idInCore) {
             // Skip over ourselves
             continue;
         }
+        // Choose a victim core that we will pawn our work on.
+        threadClass_t threadClass = core.localThreadContexts[i]->threadClass;
+        if (threadClass > corePolicy->maxClass()) {
```
Comment (johnousterhout) [inline src/Arachne.cc]: More leakage: corePolicy->maxClass. Let's discuss.

File: src/Arachne.h [hunk 29]
```diff
@@ -342,12 +345,15 @@ struct ThreadContext {
     /// context shall wait on this CV.
     ConditionVariable joinCV;
 
-    // Unique identifier for the core that this thread currently lives on.
-    // Used to index into global arrays with information per core.
+    /// Unique identifier for the core that this thread currently lives on.
+    /// Used to index into global arrays with information per core.
```
Comment (johnousterhout) [inline src/Arachne.h]: I don't think line 349 is needed (what it says is pretty obvious).

File: src/Arachne.h [hunk 30]
```diff
@@ -342,12 +345,15 @@ struct ThreadContext {
...
     /// This will only change if a ThreadContext is migrated when scaling down
     /// the number of cores.
     uint8_t coreId;
 
+    /// Thread class of this thread, used by the core policy.
```
Comment (johnousterhout) [inline src/Arachne.h]: Comment mostly duplicates the code. You could shorten to:

/// Passed to the CorePolicy to choose core(s) for execution.

Let's discuss.

File: src/Arachne.h [hunk 31]
```diff
@@ -505,16 +514,21 @@ random(void) {
   */
 template<typename _Callable, typename... _Args>
 ThreadId
-createThreadOnCore(uint32_t virtualCoreId, _Callable&& __f, _Args&&... __args) {
-    if (virtualCoreId >= numActiveCores) {
-        ARACHNE_LOG(VERBOSE, "createThread failure, virtualCoreId = %u, "
-                   "numActiveCores = %d\n", virtualCoreId,
+createThreadOnCore(threadClass_t threadClass, uint32_t coreId, _Callable&& __f, _Args&&... __args) {
```
Comment (johnousterhout) [inline src/Arachne.h]: You haven't changed the documentation to match the changes to the method.

File: src/Arachne.h [hunk 32]
```diff
@@ -505,16 +514,21 @@ random(void) {
...
+
+    threadCoreMapEntry* entry = corePolicy->getThreadCoreMapEntry(threadClass);
+    bool isLegalCoreId = false;
+    for (uint32_t i = 0; i < entry->numFilled; i++) {
+      if (entry->map[i] == (int) coreId)
```
Comment (johnousterhout) [inline src/Arachne.h]: I'm not sure what the purpose is to use both coreId and threadClass. Let's discuss.

File: src/CorePolicyTest.cc [hunk 33]
```diff
@@ -0,0 +1,179 @@
...
+static void limitedTimeWait(std::function<bool()> condition,
+                int numIterations) {
+    for (int i = 0; i < numIterations; i++) {
+        if (condition()) {
+            break;
+        }
+        usleep(1000);
+    }
+    // We use assert here because an infinite loop will result in TearDown
+    // not being able to complete, so we might as well terminate the tests here.
```
Comment (johnousterhout) [inline src/CorePolicyTest.cc]: I don't understand this. Let's discuss.

File: src/CorePolicyTest.cc [hunk 34]
```diff
@@ -0,0 +1,179 @@
...
+    corePolicy->addCore(5);
+    EXPECT_EQ(entry->map[0], 5);
+    corePolicy->addCore(4);
+    corePolicy->addCore(7);
+    EXPECT_EQ(entry->map[0], 5);
+    EXPECT_EQ(entry->map[1], 4);
+    EXPECT_EQ(entry->map[2], 7);
+    EXPECT_EQ(entry->numFilled, 3U);
+    delete corePolicy;
+}
```
Comment (johnousterhout) [inline src/CorePolicyTest.cc]: This is a decent start on tests, but it doesn't test several methods at all, such as coreLoadEstimator. Let's discuss.

**Review by johnousterhout — COMMENTED** [review]
This is not bad, and it looks functional. However, it also looks just a tad hasty: there are a few places where it looks like you took the quick and slightly dirty way out rather than finding a  really clean solution. Also, the documentation needs a good bit more work. Let's discuss once you've had a chance to read over my comments.

-John-

File: src/CorePolicy.h [hunk 35]
```diff
@@ -0,0 +1,110 @@
...
+#include "Logger.h"
+#include "PerfStats.h"
+#include "Common.h"
+
+/*
+ * Resizing of the threadClassCoreMap is not thread-safe and cannot be made
+ * thread-safe in a performant way, so for simplicity the threadClassCoreMap
+ * is allocated to a fixed size with MAX_THREAD_CLASSES threadClass slots, 
+ * which should be more than are ever needed.
+ */
```
Comment (johnousterhout) [inline src/CorePolicy.h]: This gives all source of details about how the variable is used, but doesn't actually say what it is. Start with that. Then, think about how much people really need to know in order to understand the definition of this variable. You don't need to explain every possible use of the variable; this documentation effectively duplicates significant chunks of code that appear elsewhere. Let's discuss what is the simplest thing you can say here that is complete enough.

File: src/CorePolicy.h [hunk 36]
```diff
@@ -0,0 +1,110 @@
...
+#define MAX_THREAD_CLASSES 64
+
+typedef uint32_t threadClass_t;
```
Comment (johnousterhout) [inline src/CorePolicy.h]: General question: does this code review also address my comments that weren't related to documentation? I see several things that I thought I commented on, but which don't appear to have changed, like the name of this class. Let's discuss.

File: src/CorePolicy.h [hunk 37]
```diff
@@ -0,0 +1,110 @@
...
+
+/** 
+  * An individual entry in the threadClassCoreMap. Each CoreList corresponds to
+  * a particular thread class.  For a particular CoreList corresponding to a 
+  * particular thread class, the set of all cores {map[i] for i < numFilled}
+  * is the set of cores on which threads of that class can run.
+  */
```
Comment (johnousterhout) [inline src/CorePolicy.h]: This comment says way more than it needs to. Let's discuss how to make it simpler. For example, you have described a very specific usage of this, but there's no reason why a CoreList couldn't be used in ways other than what you have described. Let's discuss how to make this documentation more abstract

File: src/CorePolicy.h [hunk 38]
```diff
@@ -0,0 +1,110 @@
...
+struct CoreList {
+  uint32_t numFilled;
+  int* map;
```
Comment (johnousterhout) [inline src/CorePolicy.h]: Every instance variable needs to be documented. Let's discuss what you need for these variables.

File: src/CorePolicy.h [hunk 39]
```diff
@@ -0,0 +1,110 @@
...
+};
+
+/*
+ * Arachne uses the CorePolicy class to make decisions about what threads
+ * to schedule on what cores.  All threads have a particular thread class,
+ * which is mapped to a set of cores on which threads of that class can run
+ * by the threadClassCoreMap, accessed through getCoreList.  All other
+ * class methods maintain the threadClassCoreMap as cores are added or removed.
```
Comment (johnousterhout) [inline src/CorePolicy.h]: This documentation has too much detail, which obscures the overall ideas. The goal is abstraction: describing things at a higher, simpler, level. Let's discuss.

File: src/CorePolicy.h [hunk 40]
```diff
@@ -0,0 +1,110 @@
...
+    virtual void addCore(int coreId);
+    virtual void removeCore(int coreId);
+    CoreList* getCoreList(threadClass_t threadClass);
+
+    /* 
+     * The default thread class, which all core policies must support.
+     * It is used as a default within Arachne and benchmarks when necessary.
+     * Other core policies can add more thread classes as needed.
+     */
+    threadClass_t defaultClass = 0;
```
Comment (johnousterhout) [inline src/CorePolicy.h]: Shouldn't this be a static const?

File: src/CorePolicy.cc [hunk 41]
```diff
@@ -0,0 +1,207 @@
...
+
+#include "Arachne.h"
+#include "CorePolicy.h"
+#include "PerfStats.h"
+#include "CoreArbiter/ArbiterClientShim.h"
+
+using PerfUtils::Cycles;
+
+/*
+ * Arachne will attempt to increase the number of cores if the idle core
```
Comment (johnousterhout) [inline src/CorePolicy.cc]: This is a little misleading: it isn't that "Arachne will attempt to increase"; it only does that because we ask it to. I'd rephrase:

"We will attempt to increase"...

File: src/CorePolicy.cc [hunk 42]
```diff
@@ -0,0 +1,207 @@
...
+ * fraction (computed as idlePercentage * numSharedCores) is less than this
+ * number.
+ */
+const double maxIdleCoreFraction = 0.1;
+
+/*
+ * Arachne will attempt to increase the number of cores if the load factor
```
Comment (johnousterhout) [inline src/CorePolicy.cc]: Same as above.

File: src/CorePolicy.cc [hunk 43]
```diff
@@ -0,0 +1,207 @@
...
+ * increases beyond this threshold.
+ */
+const double loadFactorThreshold = 1.0;
+
+/*
+ * Save the core fraction at which we ramped up based on load factor, so we
+ * can decide whether to ramp down.  Allocated in bootstrapLoadEstimator.
```
Comment (johnousterhout) [inline src/CorePolicy.cc]: Why isn't this a std::vector?

File: src/CorePolicy.cc [hunk 44]
```diff
@@ -0,0 +1,207 @@
...
+        }
+    }
+    return minCoreId;
+}
+
+/*
+ * Update threadClassCoreMap when a new core is added.  Takes in the coreId
+ * of the new core.  Arachne must guarantee that this function is not called
+ * multiple times simultaneously, or at the same time as removeCore.  This is
+ * currently enforced by the coreChangeMutex.
```
Comment (johnousterhout) [inline src/CorePolicy.cc]: This is too much detail. For example, updating threadClassCoreMap is an implementation detail; callers should not need to know that this variable even exists. In addition, I don't think coreChangeMutex needs to be mentioned here.

Also, use standard Doxygen syntax for documenting the coreId parameter.

**Review by johnousterhout — COMMENTED** [review]
I took a look at a few comments for CorePolicy.*, and I don't think you are currently thinking about documentation in the right way. Take a look at my comments, then let's discuss.

-John-

File: src/CorePolicy.cc [hunk 45]
```diff
@@ -0,0 +1,247 @@
...
+
+/*
+ * We will attempt to increase the number of cores if the load factor
+ * increases beyond this threshold.
+ */
+const double loadFactorThreshold = 1.0;
+
+/*
+ * Save the core fraction at which we ramped up based on load factor, so we
+ * can decide whether to ramp down.  Allocated in bootstrapLoadEstimator.
```
Comment (johnousterhout) [inline src/CorePolicy.cc]: Not enough information. This is an array, right? What are the indices?

Also, I'm not sure that "core fraction" is quite specific enough; is that the same as "idle core fraction"? If so, I would use the same term.

File: src/CorePolicy.cc [hunk 46]
```diff
@@ -0,0 +1,247 @@
...
+            minCoreId = coreId;
+        }
+    }
+    return minCoreId;
+}
+
+/*
+ * Update threadClassCoreMap when a new core is added.  Callers must guarantee
+ * that this function is not called multiple times simultaneously, or at the
+ * same time as removeCore.
```
Comment (johnousterhout) [inline src/CorePolicy.cc]: Is this guarantee either necessary or sufficient? If not, I would eliminate it from the interface description (it might be simpler for CorePolicies to do their own synchronization than to clutter up the interface with  this requirement). Let's discuss.

File: src/CorePolicy.cc [hunk 47]
```diff
@@ -0,0 +1,247 @@
...
+ *
+ * This function will also bootstrap the coreLoadEstimator as soon as it has
```
Comment (johnousterhout) [inline src/CorePolicy.cc]: The sentence is also implementation, not interface (e.g. callers don't need to know this).

File: src/CorePolicy.cc [hunk 48]
```diff
@@ -0,0 +1,247 @@
...
+  entry->map[entry->numFilled] = coreId;
+  entry->numFilled++;
+  if (!loadEstimatorRunning && !Arachne::disableLoadEstimation) {
+    loadEstimatorRunning = true;
+    runLoadEstimator();
+  }
+}
+
+/*
+ * Update threadClassCoreMap when a new core is removed.  Callers must
```
Comment (johnousterhout) [inline src/CorePolicy.cc]: Same comment as for addCore

File: src/CorePolicy.cc [hunk 49]
```diff
@@ -0,0 +1,247 @@
...
+  for (uint32_t i = 0; i < entry->numFilled; i++) {
+    if (entry->map[i] == coreId) {
+      entry->map[i] = saveCoreId;
+      return;
+    }
+  }
+}
+
+/*
+ * Return the CoreList for a particular threadClass.  This function is not
```
Comment (johnousterhout) [inline src/CorePolicy.cc]: Again, the first sentence is at too low a level (e.g. what does the CoreList indicate?). Let's discuss.

Also, I think you need to be more clear about exactly what "stale data" can mean: for example, can it mean returning a null pointer?

File: src/CorePolicy.cc [hunk 50]
```diff
@@ -0,0 +1,247 @@
...
+ * synchronized with addCore and removeCore and can return stale data.
+ * Callers of this function must ensure that use of stale data is safe.
+ *
+ * \param threadClass
+ *     The threadClass whose CoreList will be returned.
```
Comment (johnousterhout) [inline src/CorePolicy.cc]: Another comment that is at too low a level of detail.

What is the big idea behind this method?  E.g., what does the CorePolicy need to know in order to return the "right" CoreList?

File: src/CorePolicy.cc [hunk 51]
```diff
@@ -0,0 +1,247 @@
...
+ */
+CoreList* CorePolicy::getCoreList(ThreadClass threadClass) {
+  return threadClassCoreMap[threadClass];
+}
+
+/*
+ * Bootstrap the core load estimator thread.  Called from addCore.
+ */
+void CorePolicy::runLoadEstimator() {
```
Comment (johnousterhout) [inline src/CorePolicy.cc]: This method isn't needed; just move its contents to the (one) place where you call the method.

File: src/CorePolicy.cc [hunk 52]
```diff
@@ -0,0 +1,247 @@
...
+ */
+void coreLoadEstimator(CorePolicy* corePolicy) {
+    utilizationThresholds = new double[Arachne::maxNumCores];
+    Arachne::PerfStats previousStats;
+    Arachne::PerfStats::collectStats(&previousStats);
+
+    // Each loop cycle makes measurements and uses them to evaluate whether
+    // to increment the core count, decrement it, or do nothing.
+    for (Arachne::PerfStats currentStats; ; previousStats = currentStats) {
+        Arachne::sleep(MEASUREMENT_PERIOD);
```
Comment (johnousterhout) [inline src/CorePolicy.cc]: Is there some way of cleaning up so that this method will exit? If not, there should be. Let's discuss.

File: src/CorePolicy.cc [hunk 53]
```diff
@@ -0,0 +1,247 @@
...
+        if (totalUtilizedCores < utilizationThresholds[curActiveCores - 1]
+                - idleCoreFractionHysteresis &&
+                averageNumSlotsUsed < SLOT_OCCUPANCY_THRESHOLD) {
+            Arachne::decrementCoreCount();
+        }
+    }
+}
+
+/*
+ * Block a core in the kernel.
```
Comment (johnousterhout) [inline src/CorePolicy.cc]: This is implementation, not abstraction. Let's discuss.

File: src/CorePolicy.cc [hunk 54]
```diff
@@ -0,0 +1,247 @@
...
+ *
+ * \param coreId
+ *     The coreId of the core that will block
+ */
+void CoreBlocker::blockCore(int coreId) {
```
Comment (johnousterhout) [inline src/CorePolicy.cc]: I'm not sure that "blockCore" is the best name for this; that feels like it describes the implementation more than the abstraction. Let's discuss better possible names.

File: src/CorePolicy.cc [hunk 55]
```diff
@@ -0,0 +1,247 @@
...
+    std::mutex cvMutex;
+    std::unique_lock<std::mutex> lk(cvMutex);
+    std::condition_variable* cv = cvArray[coreId];
+    isSleepingArray[coreId] = true;
+    cv->wait(lk);
+    isSleepingArray[coreId] = false;
+}
+
+/*
+ * Unblock a core that is currently blocking.
```
Comment (johnousterhout) [inline src/CorePolicy.cc]: What if the given core is not currently blocked?

File: src/Arachne.cc [hunk 56]
```diff
@@ -220,6 +171,13 @@ bool useCoreArbiter = true;
 
 CoreArbiterClient* coreArbiter = NULL;
 
+/*
+ *  The CorePolicy that Arachne will use.  It is loaded in the init
+ *  function.
```
Comment (johnousterhout) [inline src/Arachne.cc]: I think you can skip the second sentence. If people really care about this, they can search for uses of the name (don't want to tie down the implementation unnecessarily).

File: src/Arachne.cc [hunk 57]
```diff
@@ -901,6 +858,9 @@ ThreadContext::initializeStack() {
  *     --stackSize
  *        The size of each user stack.
  *
+ * \param initCorePolicy
```
Comment (johnousterhout) [inline src/Arachne.cc]: Why "init" in the name? Let's discuss.

File: src/Arachne.cc [hunk 58]
```diff
@@ -1358,18 +1306,27 @@ bool makeExclusiveOnCore(bool forScaleDown) {
             virtualCoreTable[numActiveCores - 1] = core.kernelThreadId;
             break;
         }
+    if (forScaleDown) {
+      coreChangeMutex.lock();
```
Comment (johnousterhout) [inline src/Arachne.cc]: This seems a little strange: are you doing synchronization here purely because the CorePolicy needs it? If so, then I think you should let the CorePolicy do it for itself. I think I may have changed my mind on this since the last time we discussed this. Let's discuss.

File: src/CorePolicy.cc [hunk 59]
```diff
@@ -0,0 +1,247 @@
...
+
+/*
+ * Return the CoreList for a particular threadClass.  This function is not
+ * synchronized with addCore and removeCore and can return stale data.
+ * Callers of this function must ensure that use of stale data is safe.
+ *
+ * \param threadClass
+ *     The threadClass whose CoreList will be returned.
+ */
+CoreList* CorePolicy::getCoreList(ThreadClass threadClass) {
```
Comment (johnousterhout) [inline src/CorePolicy.cc]: This name reflects an implementation detail, not the abstraction. If someone sees a call to this method and has no other contextual information, how close will they be able to guess what the method actually does, and why it exists? Let's discuss.

File: src/Arachne.h [hunk 60]
```diff
@@ -76,6 +75,14 @@ extern std::vector<ThreadContext**> allThreadContexts;
 
 extern int* virtualCoreTable;
 
+extern CorePolicy* corePolicy;
+
+/*
+ * Testing-specific flag to make sure the core load estimator does not
+ * interfere with unit tests.
+ */
+extern bool disableLoadEstimation;
```
Comment (johnousterhout) [inline src/Arachne.h]: Is this still needed? Should move to CorePolicy?

**Review by johnousterhout — COMMENTED** [review]
This is definitely better, but your comments (and even some names) are still stuck at too low a level of detail, so they don't always provide clean and simple abstractions. Let's discuss after you had a chance to look through the comments.

File: src/CorePolicy.cc (hunk 14 above)
Comment (nexus49) [reply src/CorePolicy.cc]: The unsupported-filter check is a static contract assertion (the filter is rejected for every caller, authenticated or not). It doesn't leak per-tenant state, and matching the BulkDelete ordering keeps the two endpoints uniform. Leaving as is.
